File: packages/react-native/android/src/main/java/com/hotupdater/HotUpdaterConfig.kt
package com.hotupdater

/**
 * Thread-safe programmatic configuration holder for HotUpdater.
 *
 * When values are set here, they take priority over manifest metadata and
 * Android string resources. This enables brownfield/AAR setups where the RN
 * module cannot rely on the host app's manifest/strings.xml for configuration.
 */
object HotUpdaterConfig {
    @Volatile
    var fingerprintHash: String? = null

    @Volatile
    var publicKey: String? = null

    @Volatile
    var channel: String? = null

    /**
     * When set, this value is used verbatim as the storage isolation key,
     * bypassing the default `HotUpdaterPrefs_{fingerprint}_{appVersion}_{channel}`
     * composition. Use this to keep the OTA cache stable across host app version
     * bumps (e.g. key only by fingerprint + channel).
     */
    @Volatile
    var isolationKey: String? = null

    fun configure(
        fingerprintHash: String? = null,
        publicKey: String? = null,
        channel: String? = null,
        isolationKey: String? = null,
    ) {
        this.fingerprintHash = fingerprintHash
        this.publicKey = publicKey
        this.channel = channel
        this.isolationKey = isolationKey
    }

    fun clear() {
        fingerprintHash = null
        publicKey = null
        channel = null
        isolationKey = null
    }
}

File: packages/react-native/ios/HotUpdater/Internal/HotUpdaterConfig.swift
import Foundation

/// Thread-safe programmatic configuration holder for HotUpdater.
///
/// When values are set here, they take priority over `Info.plist` (iOS) and
/// manifest metadata / string resources (Android). This enables brownfield /
/// prebuilt-framework setups where the RN module cannot rely on the host app's
/// `Info.plist` for configuration.
@objcMembers
public class HotUpdaterConfig: NSObject {
    public static let shared = HotUpdaterConfig()

    private let lock = NSLock()

    private var _fingerprintHash: String?
    private var _publicKey: String?
    private var _channel: String?
    private var _isolationKey: String?

    public var fingerprintHash: String? {
        get { lock.lock(); defer { lock.unlock() }; return _fingerprintHash }
        set { lock.lock(); defer { lock.unlock() }; _fingerprintHash = newValue }
    }

    public var publicKey: String? {
        get { lock.lock(); defer { lock.unlock() }; return _publicKey }
        set { lock.lock(); defer { lock.unlock() }; _publicKey = newValue }
    }

    public var channel: String? {
        get { lock.lock(); defer { lock.unlock() }; return _channel }
        set { lock.lock(); defer { lock.unlock() }; _channel = newValue }
    }

    /// When set, used verbatim as the storage isolation key, bypassing the default
    /// `hotupdater_{fingerprint}_{appVersion}_{channel}_` composition. Use this to keep
    /// the OTA cache stable across host app version bumps (e.g. key only by fingerprint + channel).
    public var isolationKey: String? {
        get { lock.lock(); defer { lock.unlock() }; return _isolationKey }
        set { lock.lock(); defer { lock.unlock() }; _isolationKey = newValue }
    }

    @objc(configureWithFingerprintHash:publicKey:channel:isolationKey:)
    public func configureWithFingerprintHash(_ fingerprintHash: String?, publicKey: String?, channel: String?, isolationKey: String?) {
        self.fingerprintHash = fingerprintHash
        self.publicKey = publicKey
        self.channel = channel
        self.isolationKey = isolationKey
    }

    public func clear() {
        configureWithFingerprintHash(nil, publicKey: nil, channel: nil, isolationKey: nil)
    }
}
